app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java

```diff
@@ -0,0 +1,458 @@
+    private ArrayList<IndexedSong> queue;             // list of element currently saved (way better than songs to ensure only the correct occurrence of a song is modified)
+    private final ArrayList<IndexedSong> originalQueue; // copy of the queue used to allow revert of history last operation
+
+
+    public StaticPlayingQueue() {
+        songs = new ArrayList<>();
+        queue = new ArrayList<>();
+        originalQueue = new ArrayList<>();
+        shuffleMode = SHUFFLE_MODE_NONE;
+        currentPosition = INVALID_POSITION;
+        songsIsStale = false;
+    }
+
+    public StaticPlayingQueue(ArrayList<IndexedSong> restoreQueue, ArrayList<IndexedSong> restoreOriginalQueue, int restoredPosition, int shuffleMode) {
+        this.queue = new ArrayList<>(restoreQueue);
+        this.originalQueue = new ArrayList<>(restoreOriginalQueue);
+        this.shuffleMode = shuffleMode;
+
+        currentPosition = restoredPosition;
+
+        songs = new ArrayList<>();
+        songsIsStale = true;
+        resetSongs();
+    }
+
+    public void restoreMode(int shuffleMode, int repeatMode) {
+        this.shuffleMode = shuffleMode;
+        this.repeatMode = repeatMode;
+    }
+
+    /* -------------------- queue modification (add, remove, move, ...) -------------------- */
+
+    // add song at the end of both list
+    public void add(Song song) {
+        queue.add(new IndexedSong((Song)song, queue.size()));
+        originalQueue.add(new IndexedSong((Song)song, originalQueue.size()));
+
+        songsIsStale = true;
+    }
+
+    // add list of song at the end of both list
+    public void addAll(@NonNull List<Song> songs) {
+        for (Song song : songs) {
+            add(song);
+        }
+    }
+
+    // add song after and including position, numbering need to be redone for every song after this position (+1)
+    public void addAfter(int position, Song song) {
+        if (position >= queue.size()) {
+            position = queue.size() - 1;
+        }
+        int previousPosition = queue.get(position).index + 1;
+        position = position + 1;
+
+        originalQueue.add(previousPosition, new IndexedSong(song, previousPosition));
+        queue.add(position, new IndexedSong(song, previousPosition));
+        for (int i = 0; i < queue.size(); i++) {
+            originalQueue.get(i).index = i;
+
+            if (i != position && queue.get(i).index >= previousPosition) {
+                queue.get(i).index = queue.get(i).index + 1;
+            }
+        }
+
+        if (position < this.currentPosition) {
+            this.currentPosition++;
+        }
+
+        songsIsStale = true;
+    }
+
+    public void addSongBackTo(int position, IndexedSong song) {
+        int previousPosition = song.index;
+
+        originalQueue.add(previousPosition, new IndexedSong((Song)song, previousPosition));
```

Comment:
`new IndexedSong(song, previousPosition)` can be extracted to a variable as it is reused twice